### PR TITLE
Add GraphQL variables

### PR DESCRIPTION
**Pull request by joelvh**
This PR allows specifying GraphQL variables to the request. Keeps backward-compatibility.

### CODE REVIEW

**Comment by joelvh** [comment]
@lpinca not sure if you're the primary guy, but wanted to see if you know why the coverage test is failing?

**Comment by lpinca** [comment]
probably because the additional if branch you added is not covered by a test.

**Comment by alexandresaiz** [comment]
@lpinca hey luigi. cana you decide on this PR?

**Comment by lpinca** [comment]
I will not review it until checks are green.

**Comment by lpinca** [comment]
FWIW I'm :+1: on the feature. I suggested it when GraphQL support was added. See https://github.com/MONEI/Shopify-api-node/pull/282#discussion_r293100938.

**Comment by joelvh** [comment]
Got tests passing.

**Comment by joelvh** [comment]
@lpinca merge?

**Comment by lpinca** [comment]
Thanks, I will take a look when a I can.

File: index.js [hunk 1]
```diff
@@ -167,7 +167,7 @@ Shopify.prototype.updateGqlLimits = function updateGqlLimits(throttle) {
  * @return {Promise}
  * @public
  */
-Shopify.prototype.graphql = function graphql(data) {
+Shopify.prototype.graphql = function graphql (data, variables) {
```
Comment (lpinca) [inline index.js]:
```suggestion
Shopify.prototype.graphql = function graphql(data, variables) {
```

File: index.js [hunk 2]
```diff
@@ -176,16 +176,23 @@ Shopify.prototype.graphql = function graphql(data) {
 
   path += '/graphql.json';
 
+  const json = typeof variables !== 'undefined' && variables !== null;
```
Comment (lpinca) [inline index.js]:
```suggestion
  const json = variables !== undefined && variables !== null;
```

File: index.js [hunk 3]
```diff
@@ -176,16 +176,23 @@ Shopify.prototype.graphql = function graphql(data) {
...
+    variables
+  }) : data;
+
   const url = assign({ path }, this.baseUrl);
   const options = assign({
     timeout: this.options.timeout,
     retries: 0,
     method: 'POST',
-    body: data
+    responseType: 'json',
```
Comment (lpinca) [inline index.js]: I think this option does not exist on got@8.3.1 It should be `json` and set only when the `json` variable above is `true`.

File: index.js [hunk 4]
```diff
@@ -176,16 +176,23 @@ Shopify.prototype.graphql = function graphql(data) {
...
+    body
   }, url);
 
   options.headers['User-Agent'] = `${pkg.name}/${pkg.version}`;
-  options.headers['Content-Type'] = 'application/graphql';
+  options.headers['Content-Type'] = json ? 'application/json' : 'application/graphql';
```
Comment (lpinca) [inline index.js]: Set this to `'application/graphql'` only when `json` is `false`. The `json` option will take care of adding `'application/json'` when it is set.

File: index.js [hunk 5]
```diff
@@ -206,6 +213,17 @@ Shopify.prototype.graphql = function graphql(data) {
       this.updateGqlLimits(res.body.extensions.cost.throttleStatus);
     }
 
+    if (res.body.errors) {
+      const first = res.body.errors[0];
+      const err = new Error(first.message);
```
Comment (lpinca) [inline index.js]: With the REST API this is one of https://github.com/sindresorhus/got/tree/v8.3.1#errors but I don't any of those classes makes sense here. I think a generic `Error` is ok. What do you think?

File: index.js [hunk 6]
```diff
@@ -206,6 +213,17 @@ Shopify.prototype.graphql = function graphql(data) {
...
+      err.locations = first.locations;
+      err.path = first.path;
+      err.extensions = first.extensions;
+      err.errors = res.body.errors;
```
Comment (lpinca) [inline index.js]:
```suggestion
      err.response = res
```

Comment (lpinca) [reply index.js]: For consistency with the REST API.

File: index.js (hunk 3 above)
Comment (joelvh) [reply index.js]: Isn't the response always JSON?

Comment (lpinca) [reply index.js]: Yes but afaik this option is not available on got@8.3.1 so it probably has not effect here. Makes sense?

Comment (lpinca) [reply index.js]: See, https://github.com/sindresorhus/got/tree/v8.3.2 there is no `responseType` option. Anyway on second look we try to `JSON.parse()` so we don't need this option anyway or we will end parsing twice. I think we should completely remove this option.

File: index.js (hunk 4 above)
Comment (lpinca) [reply index.js]: No longer valid

File: test/shopify.test.js [hunk 7]
```diff
@@ -403,6 +403,64 @@ describe('Shopify', () => {
       });
     });
 
+    it('returns an Error with GraphQL info when the request fails', () => {
```
Comment (lpinca) [inline test/shopify.test.js]:
```suggestion
    it('returns an Error with GraphQL info if the request fails (1/2)', () => {
```

File: test/shopify.test.js [hunk 8]
```diff
@@ -403,6 +403,64 @@ describe('Shopify', () => {
...
+      }, err => {
+        expect(err).to.be.an.instanceof(Error);
+        expect(err.message).to.equal(message);
+        expect(err.locations).to.deep.equal(locations);
+        expect(err.path).to.equal(path);
+        expect(err.extensions).to.deep.equal(extensions);
+      });
+    });
+
+    it('returns an Error with GraphQL info when the request fails with variables', () => {
```
Comment (lpinca) [inline test/shopify.test.js]:
```suggestion
    it('returns an Error with GraphQL info if the request fails (2/2)', () => {
```

File: test/shopify.test.js [hunk 9]
```diff
@@ -540,7 +598,7 @@ describe('Shopify', () => {
         });
     });
 
-    it('returns a valid response when using graphql endpoint', () => {
+    it('returns a valid response when using graphql endpoint without variables', () => {
```
Comment (lpinca) [inline test/shopify.test.js]:
```suggestion
    it('returns a valid response when using graphql endpoint (1/2)', () => {
```

File: test/shopify.test.js [hunk 10]
```diff
@@ -553,6 +611,19 @@ describe('Shopify', () => {
         .then(res => expect(res).to.deep.equal(response.data));
     });
 
+    it('returns a valid response when using graphql endpoint with variables', () => {
```
Comment (lpinca) [inline test/shopify.test.js]:
```suggestion
    it('returns a valid response when using graphql endpoint (2/2)', () => {
```

File: index.js (hunk 3 above)
Comment (joelvh) [reply index.js]: That’s what I ended up doing. Will push the update shortly.

File: index.js (hunk 5 above)
Comment (joelvh) [reply index.js]: Yeah, generic error is simplest for now I think.

File: index.js (hunk 3 above)
Comment (alexandresaiz) [reply index.js]: @joelvh any way I can DM ?

Comment (joelvh) [reply index.js]: hi @alexandresaiz sent you an email.

**Comment by joelvh** [comment]
@lpinca resolved conflicts - your updates should all be there.

File: index.js (hunk 6 above)
Comment (lpinca) [reply index.js]: It seems this regressed :)

File: test/shopify.test.js (hunk 8 above)
Comment (lpinca) [reply test/shopify.test.js]: It seems these have not been applied?

**Comment by joelvh** [comment]
Now they should be there.

**Comment by lpinca** [comment]
Thank you.

**Comment by joelvh** [comment]
Yeah, something happened with the merge conflicts :-/ All good now. Thanks for approving!

**Comment by agentlewis** [comment]
Just want to express my gratitude for this code - exactly what I needed! It will tidy up my string literals and also enable me to use types in my code! 🙏

**Pull request merged**